cpp/src/plasma/tf_plasma_op.cc

```diff
@@ -0,0 +1,383 @@
+                                             &plasma_manager_socket_name_));
+    mutex_lock lock(mu_);
+    if (!connected_) {
+      VLOG(1) << "Connecting to Plasma...";
+      ARROW_CHECK_OK(client_.Connect(plasma_store_socket_name_,
+                                     plasma_manager_socket_name_,
+                                     PLASMA_DEFAULT_RELEASE_DELAY));
+      VLOG(1) << "Connected!";
+      connected_ = true;
+    }
+  }
+
+  ~PlasmaToTensorOp() override {
+    {
+      mutex_lock lock(mu_);
+      ARROW_CHECK_OK(client_.Disconnect());
+    }
+    {
+      mutex_lock lock(h2d_stream_mu);
+      if (h2d_stream != nullptr) {
+        delete h2d_stream;
+      }
+    }
+  }
+
+  void ComputeAsync(OpKernelContext* context, DoneCallback done) override {
+    const Tensor& plasma_object_id = context->input(0);
+    CHECK_EQ(plasma_object_id.NumElements(), 1);
+    const string& plasma_object_id_str = plasma_object_id.flat<std::string>()(0);
+
+    VLOG(1) << "plasma_object_id_str: '" << plasma_object_id_str << "'";
+    const plasma::ObjectID object_id =
+        plasma::ObjectID::from_binary(plasma_object_id_str);
+
+    plasma::ObjectBuffer object_buffer;
+    {
+      mutex_lock lock(mu_);
+      // NOTE(zongheng): this is a blocking call.  We might want to (1) make
+      // Plasma asynchronous, (2) launch a thread / event here ourselves, or
+      // something like that...
+      ARROW_CHECK_OK(client_.Get(&object_id, /*num_objects=*/1,
+                                 /*timeout_ms=*/-1, &object_buffer));
+    }
+
+    const int64_t size_in_bytes = object_buffer.data->size();
+    TensorShape shape({size_in_bytes / sizeof(float)});
+    // LOG(INFO) << "Output TensorShape: " << shape.DebugString();
+    // LOG(INFO) << "size_in_bytes of the plasma object: " << size_in_bytes;
+
+    const float* plasma_data = reinterpret_cast<const float*>(object_buffer.data->data());
+    // for (int i = 0; i < size_in_bytes / sizeof(float); ++i) {
+    //   LOG(INFO) << plasma_data[i];
+    // }
+
+    Tensor* output_tensor = nullptr;
+    OP_REQUIRES_OK_ASYNC(context, context->allocate_output(0, shape, &output_tensor),
+                         done);
+
+    if (std::is_same<Device, CPUDevice>::value) {
+      std::memcpy(output_tensor->flat<float>().data(),
+                  reinterpret_cast<const float*>(object_buffer.data->data()),
+                  size_in_bytes);
+      done();
+    } else {
+#ifdef GOOGLE_CUDA
+      auto orig_stream = context->op_device_context()->stream();
+      OP_REQUIRES_ASYNC(context, orig_stream != nullptr,
+                        errors::Internal("No GPU stream available."), done);
+      auto stream_executor = orig_stream->parent();
+
+      {
+        mutex_lock l(h2d_stream_mu);
+        if (h2d_stream == nullptr) {
+          h2d_stream = new Stream(stream_executor);
+          CHECK(h2d_stream->Init().ok());
+        }
+      }
+
+      // #define PLASMA_CLIENT_DOES_NOT_EXIST 3
```

Comment:
This commented out block can all be removed.

Reply:
doing this